### PR TITLE
Enumerating partial overlaps as an iterator

### PR DESCRIPTION
**Pull request by kris-brown**
Previously, the subobjects in order of decreasing size were computed as an iterator, but the partial overlaps between ACSets was computed all at once. However, for things like maximum common subacset, we are not interested in enumerating all possible overlaps, so this PR modifies the partial overlap computation to be an iterator.

### CODE REVIEW

File: src/categorical_algebra/CSets.jl [hunk 1]
```diff
@@ -1734,6 +1734,28 @@ function Base.iterate(Sub::SubobjectIterator, state=SubobjectIteratorState())
...
+  top::ACSet
+  others::Vector{ACSet}
+  function OverlapIterator(Xs::Vector{T}) where T<:ACSet
+    t, o... = Xs[sortperm(total_parts.(Xs))]
+    new(t, o)
+  end
+end
+
+function Base.collect(itr::OverlapIterator)
+  res = []; for sp in itr push!(res, sp) end; res
```
Comment (epatters) [inline src/categorical_algebra/CSets.jl:1747]: Is this method needed? Given that the iteration interface is implemented, I would expect `Base.collect` to already do this.

File: src/categorical_algebra/CSets.jl [hunk 2]
```diff
@@ -1734,6 +1734,28 @@ function Base.iterate(Sub::SubobjectIterator, state=SubobjectIteratorState())
 end
 
 
+struct OverlapIterator
+  top::ACSet
+  others::Vector{ACSet}
+  function OverlapIterator(Xs::Vector{T}) where T<:ACSet
+    t, o... = Xs[sortperm(total_parts.(Xs))]
```
Comment (epatters) [inline src/categorical_algebra/CSets.jl:1741]: If I understand correctly,

```julia
Xs[sortperm(total_parts.(Xs))]
```

can be simplified to

```julia
sort(Xs, by=total_parts)
```

File: src/categorical_algebra/CSets.jl (hunk 1 above)
Comment (kris-brown) [reply src/categorical_algebra/CSets.jl]: ah I just had to define Base.IteratorSize for OverlapIterator!

**Review by epatters — APPROVED** [review]
Looks good, thanks Kris!

**Pull request merged**